src/cargo/util/lints.rs

```diff
@@ -217,3 +235,113 @@ pub fn check_implicit_features(
     }
     Ok(())
 }
+
+const UNUSED_OPTIONAL_DEPENDENCY: Lint = Lint {
+    name: "unused_optional_dependency",
+    desc: "unused optional dependency",
+    groups: &[],
+    default_level: LintLevel::Warn,
+    edition_lint_opts: None,
+};
+
+pub fn unused_dependencies(
+    pkg: &Package,
+    path: &Path,
+    lints: &TomlToolLints,
+    error_count: &mut usize,
+    gctx: &GlobalContext,
+) -> CargoResult<()> {
+    let edition = pkg.manifest().edition();
+    // This lint is only relevant for editions 2024 and later.
```

Comment:
nit: to get at more of the intent
```suggestion
    // Dependencies can only be used on edition 2024+
```

Reply:
This suggested change makes it more confusing to me. We can always have dependencies, right?

Maybe:

> Unused optional dependencies are always allowed prior to Edition 2024.

Reply:
I think I left off "Unused" at the start of my suggestion

The problem with that suggestion is that unused optional dependencies don't exist prior to 2024